### PR TITLE
Paper: Training machine learning models faster with PyTorch and Dask

### PR DESCRIPTION
**Pull request by stsievert**
* **Title:** Training machine learning models faster with Dask
* **Authors:** @JoeHolt and I
* **Presentation type:** Poster.

The **current draft** of this paper can be found at https://procbuild.scipy.org/download/stsievert-adadamp. If it isn't built, visit http://procbuild.scipy.org/ and trigger a build. and the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper) has more information.

*Editor*: @cbcunc

*Reviewers*: @richardliaw

### CODE REVIEW

**Comment by deniederhut** [comment]
@scoobies assign @cbcunc as editor

**Comment by scoobies** [comment]
Assigned! @cbcunc is now the editor

**Comment by cbcunc** [comment]
Editor’s initial quickie look-over:

- [x] Make sure the paper builds on procbuild
- [x] Make sure that the sections are filled in
- [x] Make sure 8 pages or less exclusive of references
- [x] No changes outside paper directory

**Comment by deniederhut** [comment]
@scoobies add @richardliaw to reviewers

**Comment by scoobies** [comment]
@richardliaw added to the reviewers list!

**Comment by deniederhut** [comment]
We've been having difficulty finding reviewers for some papers, so to make sure everyone has enough time to do their reviews, we will be bumping the deadline for the initial review back to June 17th.

**Comment by deniederhut** [comment]
Hey @richardliaw! Today is the deadline for initial reviews. Have you had a chance to review this paper?

File: papers/scott_sievert/main.rst [hunk 1]
```diff
@@ -0,0 +1,726 @@
...
+
+   \par
+
+   Various methods to \emph{adaptively} change the batch size based on model
+   performance have been proposed \cite{sievert2021improving, de2016big,
+   balles2016coupling, byrd2012}.    Of course, these methods are adaptive so
+   computing the batch size requires computation (though there are
+   workarounds~\cite{sievert2021improving, balles2016coupling}). The
+   convergence results for these adaptive methods suggest passive methods of
+   increasing the batch size \cite{sievert2021improving}.
```
Comment (richardliaw) [inline papers/scott_sievert/main.rst]: Hmm, I didn't quite understand this sentence. Should this sentence be formed as "[sic] These results suggest *that* passive methods ... are better" (or something?)

File: papers/scott_sievert/main.rst [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+
+
+.. latex::
+
+   The key feature of AdaDamp is that the number of workers grows
+   with the batch size. Then, the model update time is agnostic to the batch size
+   (provided communication is instantaneous). This has been shown empirically:
+   Goyal et al. grow the batch size (and the number of workers with it) by a
+   factor of $44$ but the time for a single model update only increases by a
+   factor of $1.13$~\cite[Sec.~5.5]{goyal2017accurate}.
```
Comment (richardliaw) [inline papers/scott_sievert/main.rst]: Maybe you want to highlight this key feature in the intro. I read up until here and didn't realize this is what the paper was about until this sentence.

File: papers/scott_sievert/main.rst [hunk 3]
```diff
@@ -0,0 +1,726 @@
...
+
+In distributed systems, this behavior—growing the amount of data with the
+number of workers—is known as "weak scaling." By contrast, "strong scaling" has
+a fixed batch size and treats the number of workers as an internal detail. Of
+course, weak scaling exhibits better scaling than strong scaling
+:cite:`qi2017paleo`.
```
Comment (richardliaw) [inline papers/scott_sievert/main.rst]: Should this go in the intro too? It reads more like context.

File: papers/scott_sievert/main.rst [hunk 4]
```diff
@@ -0,0 +1,726 @@
...
+
+Conclusion
+==========
+
+In this work, we have provided a package to train PyTorch ML models with Dask
+cluster. This package reduces the amount of time required to train a model with
+the current centralized setup. However, it can be further accelerated by
+integration with PyTorch's distributed communication package as illustrated by
+extensive simulations. In summary, the expected gains are to go from training
+requiring about 120 minutes to 45 minutes.
```
Comment (richardliaw) [inline papers/scott_sievert/main.rst]: maybe mention that this is for a specific model? 120 minutes can be considered very long or very short -- it's model dependent.

File: papers/scott_sievert/main.rst [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+Finally, let's show how the number of Dask workers affects the time required to
+complete a single epoch with a constant batch size. This simulation will use
+the ``decentralized-high`` network and has the advantage of removing any
+overhead. The results in Figure :ref:`fig:nworkers` show that the speedups
+start saturating around 128 examples/worker for the model used with a batch
+size of 512. Larger batch sizes will likely mirror this performance –
+computation is bottleneck with this model/dataset/hardware.
+
+.. figure:: figs/decentralized-moderate/training_time.png
+   :align: center
```
Comment (richardliaw) [inline papers/scott_sievert/main.rst:701]: do you incorporate any graphs regarding time vs loss in the manuscript?

I saw that there was a `epoch_score__loss/acc.pdf`, but it didn't seem to be included in the manuscript.

**Review by richardliaw — COMMENTED** [review]
Overall nice work! Tried to avoid nit-picking.

**Comment by cbcunc** [comment]
@scoobies help

**Comment by scoobies** [comment]
Hello @cbcunc, here are the things you can ask me to do:

```

# List all available commands
@scoobies help

# Show our community Code of Conduct and Guidelines
@scoobies code of conduct

# Add to this issue's reviewers list
@scoobies add @username to reviewers

# Remove from this issue's reviewers list
@scoobies remove @username from reviewers

# Assign a user as the editor of this submission
@scoobies assign @username as editor

# Remove the editor assigned to this submission
@scoobies remove editor

# Add a user to this issue's assignees list
@scoobies add assignee: @username

# Remove a user from this issue's assignees list
@scoobies remove assignee: @username

# Builds paper
@scoobies build paper

# Checks build status
@scoobies build status

# Check the references of the paper for missing DOIs
# Optionally, it can be run on a non-default branch 
@scoobies check references
@scoobies check references from custom-branch-name

# Label issue with: paper
@scoobies mark paper

# Label issue with: needs-review. Remove labels: unready, needs-more-review, pending-comment, ready
@scoobies mark needs review

# Label issue with: needs-more-review. Remove labels: unready, needs-review, pending-comment, ready
@scoobies mark needs more review

# Label issue with: pending-comment. Remove labels: unready, needs-review, needs-more-review, ready
@scoobies mark pending comment

# Label issue with: ready. Remove labels: unready, needs-review, needs-more-review, pending-comment
@scoobies mark ready

# Label issue with: unready. Remove labels: needs-review, needs-more-review, pending-comment, ready
@scoobies mark not ready

# Label issue with: does_not_build:server
@scoobies mark server fail

# Remove labels: does_not_build:server
@scoobies mark server success

```

**Comment by cbcunc** [comment]
@scoobies check references

**Comment by scoobies** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- None

MISSING DOIs

- 10.1007/s10107-012-0572-5 may be a valid DOI for title: Sample size selection in optimization methods for machine learning
- 10.1109/cvpr.2019.01138 may be a valid DOI for title: A sufficient condition for convergences of adam and rmsprop
- 10.1145/2640087.2644155 may be a valid DOI for title: Scaling Distributed Machine Learning with the Parameter Server.
- 10.1109/icdm.2016.0074 may be a valid DOI for title: HogWild++: A New Mechanism for Decentralized Asynchronous Stochastic Gradient Descent
- 10.1007/978-3-319-46128-1_50 may be a valid DOI for title: Linear convergence of gradient and proximal-gradient methods under the polyak-łojasiewicz condition

INVALID DOIs

- None
```

**Comment by cbcunc** [comment]
@stsievert See the DOI recommendations above at https://github.com/scipy-conference/scipy_proceedings/pull/614#issuecomment-866423273 to see if it helps you to flesh out your references. We need each reference to have a DOI wherever possible. Thanks.

**Comment by deniederhut** [comment]
@scoobies mark needs more review

**Comment by deniederhut** [comment]
@stsievert to give everyone more time to respond to their reviews, we've pushed back the deadline for authors' final revisions to July 1st

File: papers/scott_sievert/main.rst (hunk 1 above)
Comment (stsievert) [reply papers/scott_sievert/main.rst]: Resolved. Thanks!

(I deleted this sentence – turns out the passive methods aren't relevant).

File: papers/scott_sievert/main.rst (hunk 2 above)
Comment (stsievert) [reply papers/scott_sievert/main.rst]: Good idea. I mentioned it in the contributions section, just below the bullet points.

File: papers/scott_sievert/main.rst (hunk 5 above)
Comment (stsievert) [reply papers/scott_sievert/main.rst:701]: The graph with loss mirrors the one with accuracy; I've included them in the appendix.

**Comment by cbcunc** [comment]
@scoobies check references

**Comment by scoobies** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- None

MISSING DOIs

- 10.1007/s10107-012-0572-5 may be a valid DOI for title: Sample size selection in optimization methods for machine learning
- 10.1109/cvpr.2019.01138 may be a valid DOI for title: A sufficient condition for convergences of adam and rmsprop
- 10.1145/2640087.2644155 may be a valid DOI for title: Scaling Distributed Machine Learning with the Parameter Server.
- 10.1109/icdm.2016.0074 may be a valid DOI for title: HogWild++: A New Mechanism for Decentralized Asynchronous Stochastic Gradient Descent
- 10.1007/978-3-319-46128-1_50 may be a valid DOI for title: Linear convergence of gradient and proximal-gradient methods under the polyak-łojasiewicz condition

INVALID DOIs

- None
```

**Comment by stsievert** [comment]
@scoobies check references

**Comment by scoobies** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1137/16M1080173 is OK
- 10.1007/s10107-012-0572-5 is OK
- 10.1109/cvpr.2019.01138 is OK
- 10.1109/icdm.2016.0074 is OK
- 10.1007/978-3-319-46128-1_50 is OK

MISSING DOIs

- 10.1145/2640087.2644155 may be a valid DOI for title: Scaling Distributed Machine Learning with the Parameter Server

INVALID DOIs

- None
```

**Comment by stsievert** [comment]
Thanks for the review @richardliaw! I've addressed your comment and done another pass over the paper.

I'd appreciate another review if you're willing and able. Don't be afraid to nitpick (especially given the mistakes I caught!). I have verified https://procbuild.scipy.org/download/stsievert-adadamp is up-to-date.

**Comment by deniederhut** [comment]
Hey @richardliaw ! Do you feel that Scott's changes have addressed your comments?

**Comment by deniederhut** [comment]
@scoobies mark pending comment

**Comment by cbcunc** [comment]
@scoobies check references

**Comment by scoobies** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1137/16M1080173 is OK
- 10.1007/s10107-012-0572-5 is OK
- 10.1137/0330046 is OK
- 10.1109/cvpr.2019.01138 is OK
- 10.3929/ethz-b-000319485 is OK
- 10.14778/3415478.3415530 is OK
- 10.1109/ICDM.2016.0074 is OK
- 10.1007/978-3-319-46128-1_50 is OK

MISSING DOIs

- 10.1145/2640087.2644155 may be a valid DOI for title: Scaling Distributed Machine Learning with the Parameter Server

INVALID DOIs

- None
```

**Comment by cbcunc** [comment]
@scoobies mark ready
